### ISSUE TITLE
Slightly wrong concrete transaction data in multi-tx & multi-contract (delegatecall) scenario

### ISSUE THREAD

**Issue by b-mueller** [ConsenSys/mythril#674]
## Description

Running Mythril on a relatively complex example appears to detect the correct bug, but returns slightly wrong concrete transaction data to reach it. Here is the code:

```solidity
pragma solidity ^0.4.24;

contract EvilDelegate {

  bool xyz;

  function abc() public {
    // do nothing
    xyz = false;
  }
}

contract Wallet {

  bool private initialized = false;
  address public owner;
  EvilDelegate private _delegate;

  constructor(address _delegateAddress) public {
    _delegate = EvilDelegate(_delegateAddress);
  }
  
  // The owner can only be set once during initialization.
  function initWallet() public {
    require(initialized == false);
    initialized = true;
    owner = msg.sender;
  }

  // Dangerous! Can only be called by the owner!
  function kill() public {
      require(msg.sender == owner);
      selfdestruct(msg.sender);
  }
  
  function() public payable {
    if (msg.data.length > 0) {
      _delegate.delegatecall(msg.data);
    }
  }
}
```

Wallet is deployed on Ropsten with address [0xc2e75b81d21831c3bf6a739c3b1495c33045d258](https://ropsten.etherscan.io/address/0xc2e75b81d21831c3bf6a739c3b1495c33045d258) and the owner has been initialized to 0x0583858adfe7d1d3DC7F8447301cf7B7b056d638.

Mythril output is as follows:

```
$ $MYTH --rpc infura-ropsten -msuicide -xla 0xc2e75b81d21831c3bf6a739c3b1495c33045d258 --transaction-count 3 --verbose-report
==== Unchecked SUICIDE ====
SWC ID: 106
Type: Warning
Contract: 0xc2e75b81d21831c3bf6a739c3b1495c33045d258
Function name: kill()
PC address: 572
A reachable SUICIDE instruction was detected. The remaining Ether is sent to the caller's address.

--------------------
--------------------
DEBUGGING INFORMATION:

Transaction Sequence: {'1': {'calldata': '0x00', 'call_value': None, 'caller': '0x0000000000000000000000000000000000000001'}, '6': {'calldata': '0x3e326048', 'call_value': '0x0', 'caller': '0x0000000000000000000000000000000000000001'}, '16': {'calldata': '0x41c0e1b5', 'call_value': '0x0', 'caller': '0x0000000000000000000000000000000000000001'}}
```
## Expected behavior

Theoretically the correct call data for the 3-tx "exploitation" sequence is:

- '0x92277933' -> call `abc()` which is forwarded through the delegate call proxy and sets `initialized` to `false`
- '0x3e326048'  -> call `initWallet()` which set `owner` to `msg.sender`
- '0x41c0e1b5' -> call `kill()`

The output produced by Mythril is:

- '0x00' -> empty calldata
- '0x3e326048'  -> call `initWallet()` which set `owner` to `msg.sender`
- '0x41c0e1b5' -> call `kill()`

Might be something wrong in processing calldata passed between contracts in the delegatecall, or an issue resolving storage.

Note that without the `-l` flag no issue is detected (which is expected behavior).

## Environment

- Mythril version: Develop
- Python version: `python -V`
- OS and Version: Mac OS Mojave

**Comment by nbanmp** [comment]
I'm also getting the same bug with other types of calls as well.
```
pragma solidity ^0.4.24;

contract OtherContract {
    function abc(address _addr) public returns (address){
        return _addr;
    }
}

contract RealContract {

    address public owner;
    OtherContract private _contract;

    constructor(address _contractAddress) public {
        _contract = OtherContract(_contractAddress);
    }

    function oops(address _addr) public {
        owner = _contract.abc(_addr);
    }

    // Dangerous! Can only be called by the owner!
    function kill() public {
        require(msg.sender == owner);
        selfdestruct(msg.sender);
    }
}
```

Output is: 
```
==== Unchecked SUICIDE ====
SWC ID: 106
Type: Warning
Contract: RealContract
Function name: kill()
PC address: 698
A reachable SUICIDE instruction was detected. The remaining Ether is sent to the caller's address.

--------------------
In file: ./doesthiswork.sol:26

selfdestruct(msg.sender)

--------------------
--------------------
DEBUGGING INFORMATION:

Transaction Sequence: {'5': {'calldata': '0x3d50779d', 'call_value': '0x0', 'caller': '0x0000000000000000000000000000000000000001'}, '6': {'calldata': '0x41c0e1b5', 'call_value': '0x0', 'caller': '0x0000000000000000000000000000000000000039'}}
```
The first transaction's calldata should have the address that is passed into the call, but it does not.

**Comment by nbanmp** [comment]
I take that back, it's a completely different issue with the example I sent.

The issue in the original bug is that we do not support internal calls with a symbolic calldata offset. Sending a raw call with `msg.data` as the calldata data has a symbolic offset, the size is calldatasize, which, with a symbolic calldata, is symbolic.

**Comment by b-mueller** [comment]
> We might be able to add a special case for passing the entire calldata is through when needed though.

Interesting, yes, we should do that, since delegate call proxies are a common pattern.

**Comment by nbanmp** [comment]
https://github.com/ConsenSys/mythril-classic/pull/739 and #693 have been closed. Some good progress, but not good enough and they don't fix the issue.

**Comment by palkeo** [comment]
I'm finally able to reproduce. I found various regressions before being able to:
 - the mutation pruner will apparently prune transactions where the delegatecall caused SSTORE. This led to not being able to simulate more than one transaction. By propagating the annotations to the parent transaction this fixes it.
 - in Storage, printable_storage is emptied every time it is copied (because of a wrong variable name)
 - after fixing the previous point, the json serialization of reports crashes
 - my fix for #1176 was wrong, and currently is causes the new owner address to be overwritten by another read from the blockchain storage.

I'll try to come up with fixes for these, and then fix the actual problem. Also I think the current behavior is not so bad: better to unconstrain the input so we catch the bug, even if it's slightly wrong.
